Add content related to React component name tracking (#9097)
getsentry/sentry-docs: docs/platforms/javascript/guides/react/features/component-tracking.mdx

description: "Learn how Sentry's React SDK allows you to monitor the rendering performance of your application and its components."
---

Sentry offers the ability to track your React components to unlock additional insights in your application.

## Component Name Tracking

> You can track the names of React components in your application via a [Babel plugin](https://www.npmjs.com/package/@sentry/babel-plugin-component-annotate), which can unlock powerful workflows. For example, you can search for replays where a certain React component was clicked on. You can also see the name of the component that was clicked on in breadcrumbs, and spans in Performance. This considerably decreases the level of ambiguity regarding the specific element on the page that was interacted with, as you would normally see just the element's CSS selector (which is also subject to further ambiguity during the minification process).
> 

Reviewer: ```suggestion
You can track the names of React components in your application via a [Babel plugin](https://www.npmjs.com/package/@sentry/babel-plugin-component-annotate), which can unlock powerful workflows and decrease ambiguity. For example, you can search for replays where a specific React component was clicked on. You can also see the name of the component that was clicked on in breadcrumbs and spans on Sentry's **Performance** page. This significantly decreases the amount of ambiguity involved in figuring out which element was interacted with as compared to trying to figure it out by looking at the element's CSS selector alone (which becomes even more ambiguous after the minification process).

```

> There will be more features released down the line that will leverage component name tracking, so it is highly recommended to configure your project to use it.

Reviewer: ```suggestion
We're working to release more features that will leverage component name tracking in the future and highly recommended that you configure your project to use it.
```

### How to Install

<Alert>

Please note that your Sentry browser SDK must be at version `7.91.0` or higher before you can use these features.
Only React components in `.jsx` or `.tsx` files can be tracked.

</Alert>

> There are two different ways you can setup React component name tracking in your application:
> 
> 1. Install [Sentry's bundler plugins](https://www.npmjs.com/package/@sentry/bundler-plugin-core)
> 2. Directly install [@sentry/babel-plugin-component-annotate](https://www.npmjs.com/package/@sentry/babel-plugin-component-annotate)
> 

Reviewer: ```suggestion
There are two different ways you can set up React component name tracking in your application:

1. By installing [Sentry's bundler plugins](https://www.npmjs.com/package/@sentry/bundler-plugin-core)
2. By directly installing [@sentry/babel-plugin-component-annotate](https://www.npmjs.com/package/@sentry/babel-plugin-component-annotate)

```

> It's recommended that you use the first option, as the Sentry bundler plugins come packed with additional useful features to enrich your Sentry workflows, such as automatic sourcemap upload,
> release creation, automatic release name discovery, and release injection. However, you have the option of installing the standalone Babel plugin as well.

Reviewer: ```suggestion
We recommend that you use the first option because the Sentry bundler plugins come packed with additional useful features to enrich your Sentry workflows, including: automatic sourcemap upload,
release creation, automatic release name discovery, and release injection. 
```

> **Setup via bundler plugins**:

Reviewer: ```suggestion
**Set up using bundler plugins**:
```

Follow the instructions in the npm page for the bundler that your project uses:
- [Rollup](https://www.npmjs.com/package/@sentry/rollup-plugin)
- [Vite](https://www.npmjs.com/package/@sentry/vite-plugin)
- [Webpack](https://www.npmjs.com/package/@sentry/webpack-plugin)

> Please note that although there is a Sentry bundler plugin for **esbuild**, React component name tracking is not currently supported by it.
> 

Reviewer: ```suggestion
Please note that although there is a Sentry bundler plugin for **esbuild**, React component name tracking is currently not supported.

```

> Once you've followed the instructions to install the bundler plugin and added it to your bundler's config, enable component name tracking by setting the flag
> to enable it to `true`:

Reviewer: ```suggestion
Once you've followed the instructions to install the bundler plugin and added it to your bundler's config, enable component name tracking by setting the flag for it to `true`:
```

```javascript
// Example specific to Vite, see documentation for other bundlers
sentryVitePlugin({
    // ... other options above
    reactComponentAnnotation: { enabled: true }
}),
```

<Alert>

> **Important Note**: If you are using **Rollup** as your bundler, ensure that you place `sentryRollupPlugin` before any other plugins that transform your JSX
> 

Reviewer: ```suggestion
**If you're using Rollup as your bundler**, ensure that you place `sentryRollupPlugin` before any other plugins that transform your JSX.

```

</Alert>

> Now, you will be able to see your component names in breadcrumbs, spans, and search for replays by component names.
> See [this page](/platforms/javascript/guides/react/features/component-tracking/#component-name-tracking) for more information on searching for Replays by component name.
> 

Reviewer: ```suggestion
Now you're all set to be able to see your component names in breadcrumbs and spans, and to search for replays by component name.
```

Author: Looks good, but why remove the link here?

### How It Works

> The Babel plugin parses your application's JSX source code at build-time, and applies additional `data` attributes onto it.
> These attributes will appear on the DOM nodes of your application's built HTML, indicating the React component and file that each node is sourced from. For example, if you had a component named
> `MyAwesomeComponent` in the file `myAwesomeComponent.jsx`:

Reviewer: ```suggestion
The Babel plugin parses your application's JSX source code at build time, and applies additional `data` attributes onto it. These attributes then appear on the DOM nodes of your application's built HTML, indicating the React component and file that each node is sourced from. 

For example, if you had a component named `MyAwesomeComponent` in the file `myAwesomeComponent.jsx`:
```

```javascript
function MyAwesomeComponent() {
  return <div>This is a really cool and awesome component!</div>
}
```

> After your bundler applies the plugin and builds your project, the resulting DOM node would look like this:

Reviewer: ```suggestion
After your bundler applied the plugin and built your project, the resulting DOM node would look like this:
```

```html
<div data-sentry-component="MyAwesomeComponent" data-sentry-source-file="myAwesomeComponent.jsx">This is a really cool and awesome component!</div>
```

> The Sentry browser SDK will pick off the value from these `data` attributes and collect them when your components are interacted with.

Reviewer: ```suggestion
The Sentry browser SDK would then pick off the value from these `data` attributes and collect them when your components would be interacted with.
```
